indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -1316,22 +1304,76 @@ public ImmutableWorkerInfo apply(ZkWorker input)
   {
     return ImmutableList.copyOf(
         Collections2.transform(
-          workers,
-          new Function<ZkWorker, Worker>()
-          {
-            @Override
-            public Worker apply(ZkWorker input)
+            workers,
+            new Function<ZkWorker, Worker>()
             {
-              return input.getWorker();
+              @Override
+              public Worker apply(ZkWorker input)
+              {
+                return input.getWorker();
+              }
             }
-          }
         )
     );
   }
 
   public Collection<ImmutableWorkerInfo> getBlackListedWorkers()
   {
-    return getImmutableWorkerFromZK(blackListedWorkers);
+    synchronized (blackListedWorkers) {
+      return getImmutableWorkerFromZK(blackListedWorkers);
+    }
+  }
+
+  private boolean shouldRemoveNodeFromBlackList(ZkWorker zkWorker)
+  {
+    if (blackListedWorkers.size() > zkWorkers.size() * (config.getMaxPercentageBlacklistWorkers() / 100.0)) {
```

Comment:
I wasn't sure this would work properly either and tested that after iterator.remove() is called, a subsequent call to size() does return a decremented value